C++ API: torch::nn::BatchNorm1d

nuka137 opened the pull request:
Add torch::nn::BatchNorm1d function/module support for the C++ API.
torch::nn::BatchNorm{2,3}d will be added after this PR is merged.

Related Issue: #25883 

Reviewer: @yf225 

I would like to discuss about below items.

* Necessity of `num_batches_tracked` in `BatchNormImplBase`
  * `num_batches_tracked` is needed to calculate `momentum` when we do not feed `momentum` argument in Python API. But in C++ API, `momentum` argument has a default value.
  * `num_batches_tracked` is only used for counting up `BatchNorm1d::foward()` call. I think it is no necessary for user anymore.
* The design of `BatchNorm{1,2,3}dOptions`
  * We have already `BatchNormOptions` used for deprecated `BatchNorm` module. However, it is hard to use it for `BatchNorm{1,2,3}dOptions` because of the arguments disagreement of each modules.
  * In this PR, I introduce `BatchNormOptionsv2` template class for the `BatchNorm{1,2,3}dOptions`. But I'm not sure this design is good or not.

nuka137 committed:
C++ API: torch::nn::BatchNorm1d

yf225 commented on torch/csrc/api/src/nn/modules/batchnorm.cpp:
```diff
@@ -72,5 +78,75 @@ Tensor BatchNormImpl::pure_forward(
+Tensor BatchNorm1dImpl::forward(const Tensor& input) {
+  TORCH_CHECK(
+      input.dim() != 2 && input.dim() !=3,
+      "expected 2D or 3D input (got %dD input)", input.dim());         
```
I feel that we can add a virtual function `_check_input_dim` to the `BatchNormImplBase` class, override its implementation in `BatchNorm1dImpl`, and call this `_check_input_dim` function from `forward`, to match the Python version even better :D

nuka137 replied:
OK. I added pure virtual function `_check_input_dim` to `BatchNormImplBase` class.

yf225 commented on torch/csrc/api/src/nn/modules/batchnorm.cpp:
```diff
@@ -72,5 +78,75 @@ Tensor BatchNormImpl::pure_forward(
+  if (options.track_running_stats()) {
+    running_mean = this->register_buffer("running_mean", torch::zeros({options.num_features()}));
+    running_var = this->register_buffer("running_var", torch::ones({options.num_features()}));
+    num_batches_tracked = this->register_buffer("num_batches_tracked", torch::tensor(0, torch::dtype(torch::kLong)));
```
I think we should keep `num_batches_tracked` as `momentum` can still be null in the C++ version :D

nuka137 replied:
Thanks. I added the codes if `options.track_running_stats()` is `false`.
Do these codes match your intention?

yf225 reviewed (commented):
@nuka137 My sincere apologies for the delay and thanks so much for the investigation and the awesome work! I left some comments regarding the design choices. The scope of work is large for `BatchNorm{1,2,3}d` and thanks so much for working on it. :D

nuka137 committed:
Apply review comments

nuka137 commented on torch/csrc/api/include/torch/nn/options/batchnorm.h:
```diff
@@ -30,5 +30,34 @@ struct TORCH_API BatchNormOptions {
+
+  /// A momentum multiplier for the mean and variance.
+  /// Changing this parameter after construction __is effective__.
+  TORCH_ARG(double, momentum) = 0.1;
```
Thanks. I fixed it.

nuka137 commented on torch/csrc/api/include/torch/nn/options/batchnorm.h:
```diff
@@ -30,5 +30,34 @@ struct TORCH_API BatchNormOptions {
 };
 
+template <size_t D>
+struct BatchNormOptionsv2 {
```
Good suggestion. I use `BatchNormBaseOptions` instead of `BatchNormOptionsv2 `.

nuka137 commented on torch/csrc/api/src/nn/modules/batchnorm.cpp:
```diff
@@ -10,10 +12,14 @@
 
 BatchNormImpl::BatchNormImpl(const BatchNormOptions& options_) : options(options_) {
+  LOG(WARNING) << "torch::nn::BatchNorm module is deprecated."
+               << "Use BatchNorm{1,2,3}d instead.";
```
I agree with that. Change to `TORCH_WARN`.

nuka137 commented on torch/csrc/api/include/torch/nn/modules/batchnorm.h:
```diff
@@ -75,5 +75,58 @@ class TORCH_API BatchNormImpl : public torch::nn::Cloneable<BatchNormImpl> {
+ public:
+  using BatchNormImplBase<1, BatchNorm1dImpl>::BatchNormImplBase;
+
+  Tensor forward(const Tensor& input);
```
Thanks. I fixed it.
Also, I fixed the arguments of `F::batch_norm` too.

nuka137 commented:
@yf225 

Thanks for reviewing. I fixed all issues.
And I found that there is no `PrettyPrintBatchNorm1d` test in previous commit, so I added it in new commit.
Could you check it together?

yf225 commented on torch/csrc/api/include/torch/nn/functional/batchnorm.h:
```diff
@@ -0,0 +1,38 @@
+inline Tensor batch_norm(const Tensor& input, const Tensor& running_mean,
+                         const Tensor& running_var, const Tensor& weight,
+                         const Tensor& bias, bool training,
+                         double momentum, double eps) {
```
I believe the functional `batch_norm` can take some default values:
```python
def batch_norm(input, running_mean, running_var, weight=None, bias=None,
               training=False, momentum=0.1, eps=1e-5):
```
it would be awesome to do the same in C++ as well :D (we can use `Tensor()` as null value for `const Tensor&` type)

nuka137 replied:
Thanks. I fixed it.

yf225 commented on torch/csrc/api/include/torch/nn/functional/batchnorm.h:
```diff
@@ -0,0 +1,38 @@
+                         const Tensor& bias, bool training,
+                         double momentum, double eps) {
+  if (training) {
+    std::vector<int64_t> size = input.sizes().vec();
```
I suspect we could do
```suggestion
    auto size = input.sizes();
```
under the hood it uses `ArrayRef` as type for `size` :D

nuka137 replied:
Thanks. I followed your instruction.

yf225 commented on torch/csrc/api/include/torch/nn/functional/batchnorm.h:
```diff
@@ -0,0 +1,38 @@
+    training,
+    momentum,
+    eps,
+    torch::cuda::cudnn_is_available());
```
I think the C++ equivalent of `torch.backends.cudnn.enabled` is `at::globalContext().userEnabledCuDNN()`

nuka137 replied:
I replaced `torch.backends.cudnn.enabled` to `at::globalContext().userEnabledCuDNN()`.

yf225 commented on torch/csrc/api/include/torch/nn/functional.h:
```diff
@@ -7,3 +7,4 @@
 #include <torch/nn/functional/normalization.h>
 #include <torch/nn/functional/pooling.h>
 #include <torch/nn/functional/vision.h>
+#include <torch/nn/functional/batchnorm.h>
```
It would be awesome to put it above `#include <torch/nn/functional/distance.h>`, to sort them alphabetically :D

nuka137 replied:
OK. I followed your instruction.

yf225 commented on test/cpp/api/modules.cpp:
```diff
@@ -965,6 +965,44 @@ TEST_F(ModulesTest, BatchNormPureForward) {
+  ASSERT_FALSE(bn->weight.defined());
+  ASSERT_FALSE(bn->bias.defined());
+}
+
```
We might need a forward+backward test for `BatchNorm1d` as well, similar to other modules such as `TEST_F(ModulesTest, MultiMarginLoss)`

nuka137 replied:
OK. I added `BatchNorm1d` test to do a forward/backward test.

yf225 commented on test/cpp/api/functional.cpp:
```diff
@@ -892,3 +892,19 @@ TEST_F(FunctionalTest, Threshold) {
+  auto output = F::batch_norm(input, mean, variance, weight, bias, false,
+                              momentum, eps);
+  auto expected = (input - mean) / torch::sqrt(variance + eps);
+  ASSERT_TRUE(output.allclose(expected));
```
It would be awesome to add a `F::batch_norm(input, mean, variant)` test for only the required arguments :D

nuka137 replied:
Added `BatchNormal1dDefaultOptions` test.

yf225 reviewed (commented):
@nuka137 Thanks so much and really appreciated the awesome work! I left some comments.

Commit:
Merge branch 'master' into pr-6

yf225 commented:
It would be awesome to update the `BatchNorm1d` entry in `test/cpp_api_parity/parity-tracker.md` as well. Thanks so much for the great work!

nuka137 committed:
Apply review comments

nuka137 commented on torch/csrc/api/include/torch/nn/modules/batchnorm.h:
```diff
@@ -75,5 +75,64 @@ class TORCH_API BatchNormImpl : public torch::nn::Cloneable<BatchNormImpl> {
+// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ BatchNorm ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+/// Base class for all (dimension-specialized) batchnorm modules.
+template <size_t D, typename Derived>
```
I think template parameter `D` is still needed to print a dimension in `pretty_print` method.

nuka137 commented on torch/csrc/api/src/nn/modules/batchnorm.cpp:
```diff
@@ -72,5 +78,102 @@ Tensor BatchNormImpl::pure_forward(
+    running_mean = this->register_buffer("running_mean", Tensor());
+    running_var = this->register_buffer("running_var", Tensor());
+    num_batches_tracked = this->register_buffer("num_batches_tracked", Tensor());
+  }
```
Moved all logic to `reset` method, and deleted `reset_parameters` method and `reset_track_running_stats` method.

nuka137 commented on torch/csrc/api/src/nn/modules/batchnorm.cpp:
```diff
@@ -72,5 +78,102 @@ Tensor BatchNormImpl::pure_forward(
+  if (options.affine()) {
+    torch::nn::init::ones_(weight);
+    torch::nn::init::zeros_(bias);
+  }
```
Moved all logic to `reset` method, and deleted `reset_parameters` method and `reset_track_running_stats` method.

nuka137 commented on torch/csrc/api/src/nn/modules/batchnorm.cpp:
```diff
@@ -72,5 +78,102 @@ Tensor BatchNormImpl::pure_forward(
+void BatchNormImplBase<D, Derived>::pretty_print(std::ostream& stream) const {
+  stream << std::boolalpha
+         << "torch::nn::BatchNorm" << D << "d("
+         << "num_features=" << options.num_features() << ", "
```
I fixed it.

nuka137 commented on torch/csrc/api/include/torch/nn/options/batchnorm.h:
```diff
@@ -30,5 +30,34 @@ struct TORCH_API BatchNormOptions {
   TORCH_ARG(double, momentum) = 0.1;
 };
 
+template <size_t D>
```
I agree with that. I removed template parameter.

nuka137 commented on torch/csrc/api/include/torch/nn/options/batchnorm.h:
```diff
@@ -30,5 +30,34 @@ struct TORCH_API BatchNormOptions {
+  TORCH_ARG(bool, track_running_stats) = true;
+};
+
+using BatchNorm1dOptions = BatchNormBaseOptions<1>;
```
Thanks. I fixed it.

nuka137 commented:
@yf225 

Thanks for your instructions.
I followed them except for templatize `BatchNormImplBase`.
I think `BatchNormImplBase` is still needed to output dimension in `pretty_print`.
What do you think about it?

nuka137 commented:
@yf225 

Hi, I found that there is a change in `PrettyPrintHardtanh` test in this PR.
I believe that this is a accident from commit https://github.com/pytorch/pytorch/pull/28176/commits/ada8d540259ce7318331173100cd38dea019657f because `PrettyPrintHardtanh` test is failed now.
If you don't mind, I will fixed this in PR.

Commit:
minor fixes

Commit:
use a different design for options

Commit:
fix header include

Commit:
more renaming

Commit:
more renaming

Commit:
fix momentum

Commit:
fix momentum again

Commit:
fix functional call

Commit:
fix test

Commit:
more renaming

yf225 reviewed (approved):
@nuka137 Thanks so much for the awesome work! I took the liberty to merge `BatchNormBaseOptions` into `BatchNormOptions` because after thinking about it more I feel that forcing `F::batch_norm` to use `BatchNormBaseOptions` might not be a good idea, aesthetically speaking. Instead I renamed `features` to `num_features` and `stateful` to `track_running_stats` in `BatchNormOptions`, which is backward-compatibility breaking, but I think is better for the long term.

Commit:
fix error: comparison between signed and unsigned integer expressions

Commit:
Merge branch 'master' into pr-6
